### ISSUE TITLE
Report error 'non-first' call when repeate experiment

### ISSUE THREAD

**Issue by v3551G** [tensorflow/tensorflow#42000]
Question: The code can successfully run when using autograph for acceleration, whereas it reports 'optimizer non-first call'  error when I repeat the experiment multiple times(10 in the sample code), which confused me.
 The code logic is as follows：
--------------------------------------------
```
 for i in np.arange(10):
       run_flow(random_seed=i)

def run_flow(random_seed=0):
      optimizer = tf.keras.optimizers.Adam()
      for i in num_epoch:
           train_epoch(optimizer)

def train_epoch(optimzier):
      for i in num_batchs:
            train_step(optimizer)

@tf.function
def train_step(optimizer)
      # .....
      optimizer.apply_gradients(...)
```
----------------------------------------------
Clearly, the definition-optimizer  is outside the function-train_step.

**Comment by bhack** [comment]
It seems to me in the https://github.com/tensorflow/tensorflow/issues/27120 "famliy"

**Comment by v3551G** [comment]
> It seems to me in the #27120 "famliy"

Thank you. Still different. my code works well both in eager mode and autograph mode when not repeat experiments, I will paste the error details later.

**Comment by ravikyram** [comment]
@v3551G 

 Please, fill [issue template.](https://github.com/tensorflow/models/issues/new/choose).

Request you to share colab link or simple standalone code with supporting files to reproduce the issue in our environment.It helps us in localizing the issue faster.Thanks!

**Comment by v3551G** [comment]
> @v3551G
> 
> Please, fill [issue template.](https://github.com/tensorflow/models/issues/new/choose).
> 
> Request you to share colab link or simple standalone code with supporting files to reproduce the issue in our environment.It helps us in localizing the issue faster.Thanks!
```
# -*- coding: utf-8 -*-
import tensorflow as tf
import numpy as np

"""
Note: when add "@tf.function" in front from train_step, report error; else, successfully run two times.
Environment; tensorflow2.x python3.x
author: masterqkk
"""


def build_model():
    model = tf.keras.Sequential()
    model.add(tf.keras.layers.Dense(10, input_shape=[2]))
    model.add(tf.keras.layers.Dense(1))
    return model

def compute_loss(batch_true, batch_pred):
    losses = tf.losses.mean_squared_error(batch_true, batch_pred)
    loss = tf.reduce_mean(losses)
    return loss


@tf.function
def train_step(model, batch_input, batch_label, optimizer):
    with tf.GradientTape() as tape:
        preds = model(batch_input)
        loss = compute_loss(batch_label, preds)
    trainable_variables = model.trainable_variables
    grads = tape.gradient(loss, trainable_variables)
    optimizer.apply_gradients(grads_and_vars=zip(grads, trainable_variables))
    return loss


def train_epoch(model, data_batchs, optimizer):
    for (batch_input, batch_label) in data_batchs:
        loss = train_step(model, batch_input, batch_label, optimizer)


def train_model(model, data_batchs, optimizer):
    for i in np.arange(1):
        train_epoch(model, data_batchs, optimizer)


def load_data():
    np.random.seed(0)
    x = np.array(np.random.random(size=(10, 2)), np.float32)
    y = np.array(np.random.random(size=(10, 1)), np.float32)
    data = tf.data.Dataset.from_tensor_slices((x, y))
    data_batchs = data.batch(batch_size=5)
    return data_batchs


def run_flow():
    data_batchs = load_data()
    model = build_model()
    optimizer = tf.keras.optimizers.Adam()
    train_model(model, data_batchs, optimizer)
    tf.print('model train finished.')


for i in np.arange(2):
    run_flow()
```

**Comment by bhack** [comment]
@v3551G I still think that you are in the perimeter of https://github.com/tensorflow/tensorflow/issues/27120. 
Can you try you code with the same workaround as in https://github.com/tensorflow/tensorflow/issues/27120#issuecomment-540071844

**Comment by v3551G** [comment]
> @v3551G I still think that you are in the perimeter of #27120.
> Can you try your code with the same workaround as in [#27120 (comment)](https://github.com/tensorflow/tensorflow/issues/27120#issuecomment-540071844)

The demo for reproducing the error is pasted at my last comment， please swipe up the page.

**Comment by v3551G** [comment]
> @v3551G
> 
> Please, fill [issue template.](https://github.com/tensorflow/models/issues/new/choose).
> 
> Request you to share colab link or simple standalone code with supporting files to reproduce the issue in our environment.It helps us in localizing the issue faster.Thanks!

```
# -*- coding: utf-8 -*-
import tensorflow as tf
import numpy as np

"""
Note: when add "@tf.function" in front from train_step, report error; else, successfully run two times.
Environment; tensorflow2.x python3.x
author: masterqkk
"""


def build_model():
    model = tf.keras.Sequential()
    model.add(tf.keras.layers.Dense(10, input_shape=[2]))
    model.add(tf.keras.layers.Dense(1))
    return model

def compute_loss(batch_true, batch_pred):
    losses = tf.losses.mean_squared_error(batch_true, batch_pred)
    loss = tf.reduce_mean(losses)
    return loss


@tf.function
def train_step(model, batch_input, batch_label, optimizer):
    with tf.GradientTape() as tape:
        preds = model(batch_input)
        loss = compute_loss(batch_label, preds)
    trainable_variables = model.trainable_variables
    grads = tape.gradient(loss, trainable_variables)
    optimizer.apply_gradients(grads_and_vars=zip(grads, trainable_variables))
    return loss


def train_epoch(model, data_batchs, optimizer):
    for (batch_input, batch_label) in data_batchs:
        loss = train_step(model, batch_input, batch_label, optimizer)


def train_model(model, data_batchs, optimizer):
    for i in np.arange(1):
        train_epoch(model, data_batchs, optimizer)


def load_data():
    np.random.seed(0)
    x = np.array(np.random.random(size=(10, 2)), np.float32)
    y = np.array(np.random.random(size=(10, 1)), np.float32)
    data = tf.data.Dataset.from_tensor_slices((x, y))
    data_batchs = data.batch(batch_size=5)
    return data_batchs


def run_flow():
    data_batchs = load_data()
    model = build_model()
    optimizer = tf.keras.optimizers.Adam()
    train_model(model, data_batchs, optimizer)
    tf.print('model train finished.')


for i in np.arange(2):
    run_flow()
```

Waiting your reply.

**Comment by bhack** [comment]
I saw your code. Can you try your code with that workaround? 
Personally I tried that workaround with your code and It Is working.

**Comment by v3551G** [comment]
> I saw your code. Can you try your code with that workaround?
> Personally I tried that workaround with your code and It Is working.
do you mean commit the "@tf.function"?

**Comment by bhack** [comment]
```
# -*- coding: utf-8 -*-
import tensorflow as tf
import numpy as np

"""
Note: when add "@tf.function" in front from train_step, report error; else, successfully run two times.
Environment; tensorflow2.x python3.x
author: masterqkk
"""


def build_model():
    model = tf.keras.Sequential()
    model.add(tf.keras.layers.Dense(10, input_shape=[2]))
    model.add(tf.keras.layers.Dense(1))
    return model

def compute_loss(batch_true, batch_pred):
    losses = tf.losses.mean_squared_error(batch_true, batch_pred)
    loss = tf.reduce_mean(losses)
    return loss

def get_apply_grad_fn():
    @tf.function
    def train_step(model, batch_input, batch_label, optimizer):
        with tf.GradientTape() as tape:
            preds = model(batch_input)
            loss = compute_loss(batch_label, preds)
        trainable_variables = model.trainable_variables
        grads = tape.gradient(loss, trainable_variables)
        optimizer.apply_gradients(grads_and_vars=zip(grads, trainable_variables))
        return loss
    return train_step


def train_epoch(model, data_batchs, optimizer):
    for (batch_input, batch_label) in data_batchs:
        model_apply_grads = get_apply_grad_fn()
        loss = model_apply_grads(model, batch_input, batch_label, optimizer)


def train_model(model, data_batchs, optimizer):
    for i in np.arange(1):
        train_epoch(model, data_batchs, optimizer)


def load_data():
    np.random.seed(0)
    x = np.array(np.random.random(size=(10, 2)), np.float32)
    y = np.array(np.random.random(size=(10, 1)), np.float32)
    data = tf.data.Dataset.from_tensor_slices((x, y))
    data_batchs = data.batch(batch_size=5)
    return data_batchs


def run_flow():
    data_batchs = load_data()
    model = build_model()
    optimizer = tf.keras.optimizers.Adam()
    model_apply_grads = get_apply_grad_fn()
    train_model(model, data_batchs, optimizer)
    tf.print('model train finished.')


for i in np.arange(2):
    run_flow()
```

**Comment by nikitamaia** [comment]
Hi @v3551G, as @bhack explained the comment in #27120 to wrap the train_step is a workaround that will allow you to run multiple experiments. However, you might notice that it can have an impact on performance. The problem of not being able to repeat the experiment is actually a bug as I've noted in this comment [#36574 ](https://github.com/tensorflow/tensorflow/issues/36574#issuecomment-659040854).

**Comment by nikitamaia** [comment]
Closing this issue now since this is a known issue being tracked in #36574 and there is a temporary workaround.

**Comment by monicadsong** [comment]
As you pointed out, you are creating your optimizer outside the `tf.function`; however, you are attempting to pass a new optimizer into the same trace of `tf.function` and the optimizer creates variables the first time it applies gradients. This is why you see the error: the optimizer that is instantiated in `run_flow` attempts to create variables in `train_step` which is not allowed.

Just my two cents:

Suppose you are running a batch of 10 experiments where each experiment is called with `run_flow`:
```
for i in np.arange(10):
       run_flow(random_seed=i)
```
Currently, your code is attempting to trace the `train_step` in experiment #1 only, and reuse the trace in experiments #2 - #10. If you were successful in doing so and kept all other factors (including the data) equal in your experiments, you may find that experiment #1 takes slightly longer since tracing is an expensive computation.

If you are interested in keeping the amount of tracing in experiments #1 - #10 identical:

```
def build_model():
    model = tf.keras.Sequential()
    model.add(tf.keras.layers.Dense(10, input_shape=[2]))
    model.add(tf.keras.layers.Dense(1))
    return model

def compute_loss(batch_true, batch_pred):
    losses = tf.losses.mean_squared_error(batch_true, batch_pred)
    loss = tf.reduce_mean(losses)
    return loss

def train_step(model, batch_input, batch_label, optimizer):
    with tf.GradientTape() as tape:
        preds = model(batch_input)
        loss = compute_loss(batch_label, preds)
    trainable_variables = model.trainable_variables
    grads = tape.gradient(loss, trainable_variables)
    optimizer.apply_gradients(grads_and_vars=zip(grads, trainable_variables))
    return loss

def train_model(model, data_batchs, optimizer):
    # Create a new tf.function from train_step that will be used in this experiment only
    train_step_function = tf.function(train_step)
    num_epochs = 10
    for i in np.arange(num_epochs):
        for (batch_input, batch_label) in data_batchs:
            loss = train_step_function(model, batch_input, batch_label, optimizer)

def load_data():
    np.random.seed(0)
    x = np.array(np.random.random(size=(10, 2)), np.float32)
    y = np.array(np.random.random(size=(10, 1)), np.float32)
    data = tf.data.Dataset.from_tensor_slices((x, y))
    data_batchs = data.batch(batch_size=5)
    return data_batchs

def run_flow():
    data_batchs = load_data()
    model = build_model()
    optimizer = tf.keras.optimizers.Adam()
    train_model(model, data_batchs, optimizer)

for i in np.arange(10):
    print("Running experiment {}".format(i))
    run_flow()
```
This solution should not have a drastic impact on performance.
